cumulus/parachains/pallets/collective-content/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::{ensure, pallet_prelude::*};
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::{traits::BadOrigin, ArithmeticError::Overflow, Saturating};
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	/// The module configuration trait.
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self, I>>
+			+ IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Default lifetime for an announcement before it expires.
+		type AnnouncementLifetime: Get<BlockNumberFor<Self>>;
+
+		/// The origin to control the collective announcements.
+		type AnnouncementOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The origin to control the collective charter.
+		type CharterOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// Weight information needed for the pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T, I = ()> {
+		/// The announcement is not found.
+		MissingAnnouncement,
+		/// Number of announcements exceeds `MaxAnnouncementsCount`.
+		TooManyAnnouncements,
+		/// Cannot expire in the past.
+		InvalidExpiration,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config<I>, I: 'static = ()> {
+		/// A new charter has been set.
+		NewCharterSet { cid: OpaqueCid },
+		/// A new announcement has been made.
+		AnnouncementAnnounced { cid: OpaqueCid, expire_at: BlockNumberFor<T> },
+		/// An on-chain announcement has been removed.
+		AnnouncementRemoved { cid: OpaqueCid },
+	}
+
+	/// The collective charter.
+	#[pallet::storage]
+	#[pallet::getter(fn charter)]
+	pub type Charter<T: Config<I>, I: 'static = ()> = StorageValue<_, OpaqueCid, OptionQuery>;
+
+	/// The collective announcements.
+	#[pallet::storage]
+	#[pallet::getter(fn announcements)]
+	pub type Announcements<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Blake2_128Concat, OpaqueCid, BlockNumberFor<T>, OptionQuery>;
+
+	/// The current count of the announcements.
+	#[pallet::storage]
+	#[pallet::getter(fn announcements_count)]
+	pub type AnnouncementsCount<T: Config<I>, I: 'static = ()> = StorageValue<_, u32, ValueQuery>;
+
+	/// The closest expiration block number of an announcement.
+	#[pallet::storage]
+	pub type NextAnnouncementExpireAt<T: Config<I>, I: 'static = ()> =
+		StorageValue<_, BlockNumberFor<T>, OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Set the collective  charter.
+		///
+		/// Parameters:
+		/// - `origin`: Must be the [Config::CharterOrigin].
+		/// - `cid`: [CID](super::OpaqueCid) of the IPFS document of the collective charter.
+		///
+		/// Weight: `O(1)`.
```

Comment:
I find these weight comments pretty useless now that we have benchmarking. I would remove them from new pallets.